cmd/ci-secret-bootstrap/main.go

```diff
@@ -994,11 +1231,292 @@ func reconcileSecrets(o options, client secrets.ReadOnlyClient, prowDisabledClus
+// For bundles without explicit field lists, it discovers fields by querying GSM.
+// Returns a map of cluster name to list of Kubernetes Secret objects, and any fetch/build errors.
+func constructSecretsFromGSM(
+	ctx context.Context,
+	gsmConfig secretbootstrap.GSMConfig,
+	gsmClient gsm.SecretManagerClient,
+	gsmProjectConfig gsm.Config,
+	prowDisabledClusters sets.Set[string]) (map[string][]*coreapi.Secret, error) {
+	var errs []error
+	uniqueSecretNames := make(map[gsmSecretRef]bool)
+	discoveredFields := make(map[collectionGroupKey][]string) // track fields for collection+group pairs when `fields` stanza is empty
+
+	for _, bundle := range gsmConfig.Bundles {
+		if !bundle.SyncToCluster {
+			continue
+		}
+		for _, secretEntry := range bundle.GSMSecrets {
+			if len(secretEntry.Fields) == 0 { // if fields are not specified, discover them using GSM listing
+				key := collectionGroupKey{
+					collection: secretEntry.Collection,
+					group:      secretEntry.Group,
+				}
+
+				// Check if we've already discovered fields for this collection+group
+				if _, alreadyDiscovered := discoveredFields[key]; !alreadyDiscovered {
+					fieldNames, err := gsm.ListSecretFieldsByCollectionAndGroup(ctx, gsmClient, gsmProjectConfig, secretEntry.Collection, secretEntry.Group)
+					if err != nil {
+						errs = append(errs, fmt.Errorf("failed to list fields for collection=%s, group=%s: %w", secretEntry.Collection, secretEntry.Group, err))
+						continue
+					}
+					discoveredFields[key] = fieldNames
+					logrus.Debugf("discovered %d fields for collection=%s, group=%s", len(fieldNames), secretEntry.Collection, secretEntry.Group)
+				}
+
+				for _, fieldName := range discoveredFields[key] {
+					s := gsmSecretRef{
+						collection: secretEntry.Collection,
+						group:      secretEntry.Group,
+						field:      fieldName,
+					}
+					uniqueSecretNames[s] = true
+				}
+			} else {
+				for _, field := range secretEntry.Fields {
+					s := gsmSecretRef{
+						collection: secretEntry.Collection,
+						group:      secretEntry.Group,
+						field:      field.Name,
+					}
+					uniqueSecretNames[s] = true
+				}
+			}
+		}
+
+		if bundle.DockerConfig == nil {
+			continue
+		}
+		for _, registryEntry := range bundle.DockerConfig.Registries {
+			s := gsmSecretRef{
+				collection: registryEntry.Collection,
+				group:      registryEntry.Group,
+				field:      registryEntry.AuthField,
+			}
+			uniqueSecretNames[s] = true
+
+			if registryEntry.EmailField != "" {
+				s := gsmSecretRef{
+					collection: registryEntry.Collection,
+					group:      registryEntry.Group,
+					field:      registryEntry.EmailField,
+				}
+				uniqueSecretNames[s] = true
+			}
+		}
+	}
+
+	fetchedGsmSecretsMap := make(map[gsmSecretRef]fetchedSecret)
+	mapLock := sync.Mutex{}
+	errChan := make(chan error, len(uniqueSecretNames))
+	wg := &sync.WaitGroup{}
+
+	for secretRef := range uniqueSecretNames {
+		wg.Add(1)
+		secretRef := secretRef
```

Comment:
This shouldn't be needed anymore since Golang `1.2x` (don't remember the exact version though).